### PR TITLE
[DS-2701] REST API

### PR DESCRIPTION
**Pull request by peterdietz**
Refactoring the REST API to use the DSpace Services API refactor. REST now compiles. It would be helpful for another REST expert to review these changes with a fine-tooth comb, as this felt like a massive set of changes, that undoubtedly some changes may have occurred.

Some REST changes are that things might use a UUID instead of their legacyID.

Also, tomcat doesn't load the webapp, so, I haven't verified functionality.

Tomcat says this:
Failed to startup the DSpace Service Manager: failure starting up spring service manager: null
Failure during filter init: Failed to startup the DSpace Service Manager: failure starting up spring service manager: null:java.lang.RuntimeException: Failed to startup the DSpace Service Manager: failure starting up spring service manager: null
21-Aug-2015 18:13:43.851 SEVERE [localhost-startStop-1] org.apache.catalina.core.StandardContext.startInternal Error listenerStart
21-Aug-2015 18:13:43.862 SEVERE [localhost-startStop-1] org.apache.catalina.core.StandardContext.startInternal Context [/rest] startup failed due to previous errors

### CODE REVIEW

**Comment by tdonohue** [comment]
Hi @KevinVdV, we're assigning this to you for review / inclusion in the Services API branch.  It's ready for review & compiles.  Peter will create any additional improvements/enhancements as separate PRs.

File: dspace-rest/src/main/java/org/dspace/rest/BitstreamResource.java [hunk 1]
```diff
@@ -691,26 +686,18 @@ public Response deleteBitstream(@PathParam("bitstream_id") Integer bitstreamId,
...
-            boolean found = false;
-            List<org.dspace.authorize.ResourcePolicy> policies =  AuthorizeManager.getPolicies(context, dspaceBitstream);
-            for(org.dspace.authorize.ResourcePolicy policy : policies) {
-                if(policy.getID() == policyId) {
-                    found = true;
-                    break;
-                }
-            }
+            org.dspace.authorize.ResourcePolicy resourcePolicy = resourcePolicyService.find(context, policyId);
+            if(resourcePolicy.getdSpaceObject() == dspaceBitstream && authorizeService.authorizeActionBoolean(context, dspaceBitstream, org.dspace.core.Constants.REMOVE)) {
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/BitstreamResource.java]: This type of compare won't work as you are comparing objects & not primitive types.

File: dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java [hunk 2]
```diff
@@ -457,57 +450,56 @@ public Bitstream addItemBitstream(@PathParam("item_id") Integer itemId, InputStr
...
-                    List<org.dspace.authorize.ResourcePolicy> bitstreamsPolicies = dspaceBundle.getBitstreamPolicies();
+                    List<org.dspace.authorize.ResourcePolicy> bitstreamsPolicies = bundleService.getBitstreamPolicies(context, dspaceBundle);
 
                     // Remove default bitstream policies
                     List<org.dspace.authorize.ResourcePolicy> policiesToRemove = new ArrayList<org.dspace.authorize.ResourcePolicy>();
-                    for (org.dspace.authorize.ResourcePolicy policy : bitstreamsPolicies)
-                    {
-                        if (policy.getResourceID() == dspaceBitstream.getID())
+                    for (org.dspace.authorize.ResourcePolicy policy : bitstreamsPolicies) {
+                        if (policy.getdSpaceObject() == dspaceBitstream)
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java]: This type of compare won't work as you are comparing objects & not primitive types.

File: dspace-rest/src/main/java/org/dspace/rest/CollectionsResource.java [hunk 3]
```diff
@@ -439,14 +445,14 @@ public Response updateCollection(@PathParam("collection_id") Integer collectionI
...
-            dspaceCollection.setLicense(collection.getLicense());
+            collectionService.setMetadata(context, dspaceCollection, "name", collection.getName());
+            collectionService.setMetadata(context, dspaceCollection, "license", collection.getLicense());
+
             // dspaceCollection.setLogo(collection.getLogo()); // TODO Add this option.
-            dspaceCollection.setMetadata(org.dspace.content.Collection.COPYRIGHT_TEXT, collection.getCopyrightText());
-            dspaceCollection.setMetadata(org.dspace.content.Collection.INTRODUCTORY_TEXT, collection.getIntroductoryText());
-            dspaceCollection.setMetadata(org.dspace.content.Collection.SHORT_DESCRIPTION, collection.getShortDescription());
-            dspaceCollection.setMetadata(org.dspace.content.Collection.SIDEBAR_TEXT, collection.getSidebarText());
-            dspaceCollection.update();
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/CollectionsResource.java]: This update has been removed ?

File: dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java [hunk 4]
```diff
@@ -488,16 +486,15 @@ public Community createCommunity(Community community, @QueryParam("userIP") Stri
...
+            org.dspace.content.Community dspaceCommunity = communityService.create(null, context);
             writeStats(dspaceCommunity, UsageEvent.Action.CREATE, user_ip, user_agent, xforwardedfor,
                     headers, request, context);
 
-            dspaceCommunity.setMetadata("name", community.getName());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.COPYRIGHT_TEXT, community.getCopyrightText());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.INTRODUCTORY_TEXT, community.getIntroductoryText());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SHORT_DESCRIPTION, community.getShortDescription());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SIDEBAR_TEXT, community.getSidebarText());
-            dspaceCommunity.update();
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java]: This update has been removed ?

File: dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java [hunk 5]
```diff
@@ -565,17 +562,15 @@ public Collection addCommunityCollection(@PathParam("community_id") Integer comm
...
+            collectionService.setMetadata(context, dspaceCollection, "license", collection.getLicense());
             // dspaceCollection.setLogo(collection.getLogo()); // TODO Add this option.
-            dspaceCollection.setMetadata("name", collection.getName());
-            dspaceCollection.setMetadata(org.dspace.content.Collection.COPYRIGHT_TEXT, collection.getCopyrightText());
-            dspaceCollection.setMetadata(org.dspace.content.Collection.INTRODUCTORY_TEXT, collection.getIntroductoryText());
-            dspaceCollection.setMetadata(org.dspace.content.Collection.SHORT_DESCRIPTION, collection.getShortDescription());
-            dspaceCollection.setMetadata(org.dspace.content.Collection.SIDEBAR_TEXT, collection.getSidebarText());
-            dspaceCollection.setLicense(collection.getLicense());
-            dspaceCollection.update();
-            dspaceCommunity.update();
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java]: These updates have been removed ?

File: dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java [hunk 6]
```diff
@@ -650,14 +645,12 @@ public Community addCommunityCommunity(@PathParam("community_id") Integer commun
...
                     headers, request, context);
 
-            org.dspace.content.Community dspaceCommunity = org.dspace.content.Community.create(dspaceParentCommunity, context);
-            dspaceCommunity.setMetadata("name", community.getName());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.COPYRIGHT_TEXT, community.getCopyrightText());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.INTRODUCTORY_TEXT, community.getIntroductoryText());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SHORT_DESCRIPTION, community.getShortDescription());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SIDEBAR_TEXT, community.getSidebarText());
-            dspaceCommunity.update();
-            dspaceParentCommunity.update();
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java]: These updates have been removed ?

File: dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java [hunk 7]
```diff
@@ -729,12 +722,11 @@ public Response updateCommunity(@PathParam("community_id") Integer communityId,
                     headers, request, context);
 
             // dspaceCommunity.setLogo(arg0); // TODO Add this option.
-            dspaceCommunity.setMetadata("name", community.getName());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.COPYRIGHT_TEXT, community.getCopyrightText());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.INTRODUCTORY_TEXT, community.getIntroductoryText());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SHORT_DESCRIPTION, community.getShortDescription());
-            dspaceCommunity.setMetadata(org.dspace.content.Community.SIDEBAR_TEXT, community.getSidebarText());
-            dspaceCommunity.update();
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/CommunitiesResource.java]: This update has been removed ?

File: dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java [hunk 8]
```diff
@@ -379,10 +377,9 @@ public Response addItemMetadata(@PathParam("item_id") Integer itemId, List<org.d
                 String data[] = mySplit(entry.getKey()); // Done by my split, because of java split was not function.
                 if ((data.length >= 2) && (data.length <= 3))
                 {
-                    dspaceItem.addMetadata(data[0], data[1], data[2], entry.getLanguage(), entry.getValue());
+                    itemService.addMetadata(context, dspaceItem, data[0], data[1], data[2], entry.getLanguage(), entry.getValue());
                 }
             }
-            dspaceItem.update();
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java]: This update has been removed ?

File: dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java [hunk 9]
```diff
@@ -457,57 +450,56 @@ public Bitstream addItemBitstream(@PathParam("item_id") Integer itemId, InputStr
...
+
+                dspaceBitstream.setName(context, name);
             }
             if (description != null)
             {
-                dspaceBitstream.setDescription(description);
+                dspaceBitstream.setDescription(context, description);
             }
 
-            dspaceBitstream.update();
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java]: This update has been removed ?

File: dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java [hunk 10]
```diff
@@ -544,13 +534,12 @@ public Bitstream addItemBitstream(@PathParam("item_id") Integer itemId, InputStr
                         dspacePolicy.setStartDate(date);
                     }
 
-                    dspacePolicy.update();
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java]: This update has been removed ?

File: dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java [hunk 11]
```diff
@@ -872,11 +849,11 @@ else if (!AuthorizeServiceImpl.authorizeActionBoolean(context, bitstream, org.ds
...
             for (org.dspace.content.Bundle bundle : item.getBundles())
             {
-                for (org.dspace.content.Bitstream bit : bundle.getBitstreams())
+                for (org.dspace.content.BundleBitstream bit : bundle.getBitstreams())
                 {
-                    if (bit == bitstream)
+                    if (bit.getBitstream() == bitstream)
                     {
-                        bundle.removeBitstream(bitstream);
+                        bundleService.removeBitstream(context, bundle, bitstream);
```
Comment (KevinVdV) [inline dspace-rest/src/main/java/org/dspace/rest/ItemsResource.java]: Removing a bitstream from a bundle while looping will cause ConcurrentModificationException, best to use an iterator & remove the bitstream from the iterator prior to removing from the bundle.

**Comment by KevinVdV** [comment]
Did a quick review & was able to find a few things that require fixing. Once there are fixed I would recommend a quick rebase (do NOT merge, rebase is cleaner) on the current DS-2701-service-api branch.
There is a conflict in the "ItemService" interface were 2 different methods are added at the bottom. Both of them should of course be present in the merged result

When this is done I see no issues in merging this pull request (aside from the fact that the retrieval of DSpaceObject still uses identifiers but I am aware that this will be addressed later in separate JIRA tickets). So fix & rebase & I will merge it.

**Comment by peterdietz** [comment]
@tdonohue @KevinVdV Sorry for the delay on this, but I have addressed the issues that Kevin has brought up, and I would recommend merging this REST refactor into services.

File: dspace-rest/src/main/java/org/dspace/rest/BitstreamResource.java [hunk 12]
```diff
@@ -542,24 +558,9 @@ public Response updateBitstreamData(@PathParam("bitstream_id") Integer bitstream
...
-            originalBitstreamRow.setColumn("internal_id", bitstream.getStringColumn("internal_id"));
-            originalBitstreamRow.setColumn("size_bytes", bitstream.getLongColumn("size_bytes"));
-            bitstream.setColumn("internal_id", internal_id);
-            bitstream.setColumn("size_bytes", size_bytes);
-
-            DatabaseManager.update(context, originalBitstreamRow);
-            BitstreamStorageServiceImpl.delete(context, newBitstreamId);
-
-            context.complete();
+            UUID newBitstreamId = bitstreamStorageService.store(context, dspaceBitstream, is);
```
Comment (peterdietz) [inline dspace-rest/src/main/java/org/dspace/rest/BitstreamResource.java]: This is a good example of the difference between a UI doing logic the wrong, complicated way, and how services can simplify this.

**Pull request merged**